Move things to a Concepts module (#57)
akka/akka-microservices-samples: docs-source/docs/modules/concepts/pages/event-sourcing.adoc

:toclevels: 2

include::ROOT:partial$include.adoc[]

See an link:https://docs.microsoft.com/en-us/previous-versions/msp-n-p/jj591559(v=pandp.10)?redirectedfrom=MSDN[introduction to EventSourcing] at MSDN.

Another excellent article about “thinking in Events” is link:https://hackernoon.com/events-as-first-class-citizens-8633e8479493[Events As First-Class Citizens] by Randy Shoup. It is a short and recommended read if you’re starting developing Events based applications.

Event sourcing documentation in the link:{akka}/typed/persistence.html#introduction[Akka Reference documentation for Akka Persistence].

== Describing entities

A difference to persistence based on Create-Read-Update-Delete (CRUD) data-stores is that we don't need to map our imagined entities to a database model up-front. Instead, we model the entities and the events that affect their state.

ifdef::todo:[include something like https://www.lagomframework.com/documentation/1.6.x/scala/ES_CQRS.html and https://www.lagomframework.com/documentation/1.6.x/scala/ESAdvantage.html]

== Updating entities

To update an entity's state we use commands from the outside and events on the inside.

=== Commands

> This entity state can be changed only via commands to it. The commands are the "external" API of an entity. Commands request state changes, but they might lead to different events on the entity depending on the entities current state.

Reviewer: ```suggestion
The state of the entity can be changed only by sending commands to it. The commands are the "external" API of an entity. Commands request state changes, but they might lead to different events depending on the current state of the entity.
```

=== Events

The events represent changes of the entity's state and are the only way to change it. The entity creates events from commands, other parts of the application may listen to the created events. Other parties can't send events.

> The events get continuously persisted to the Event Journal datastore, while the entity state is kept in memory. In case of a restart the latest state gets rebuilt by replaying the events from the Event Journal.

Reviewer: ```suggestion
The events are continuously persisted to the Event Journal datastore, while the entity state is kept in memory. In case of a restart the latest state gets rebuilt by replaying the events from the Event Journal.
```
